### PR TITLE
Free Hints + Spoiler Log updates. 

### PR DESCRIPTION
**Pull request by zmarotrix**

### CODE REVIEW

File: Build.cs [hunk 1]
```diff
@@ -326,9 +326,14 @@ private void WriteSpoilerLog()
                 return;
             }
 
+
+
             if (Settings.GenerateSpoilerLog)
             {
-                MakeSpoilerLog();
+                if (!cHTMLLog.Checked)
```
Comment (ZoeyZolotova) [inline Build.cs]: Use `Settings.HTMLLog`

File: Build.cs [hunk 2]
```diff
@@ -412,6 +417,7 @@ private bool ValidateROM(string FileName)
 
         private void MakeROM(string InFile, string FileName, BackgroundWorker worker)
         {
+            if(!cLogOnly.Checked)
```
Comment (ZoeyZolotova) [inline Build.cs]: There shouldn't be a "Spoiler Log Only" checkbox. Just a checkbox for each type of output. N64, VC, SpoilerLog, SpoilerLog (HTML)

Comment (ZoeyZolotova) [inline Build.cs]: Please use curly brackets. This is too confusing to read and looks like a mistake.

File: Build.cs [hunk 3]
```diff
@@ -462,15 +468,18 @@ private void MakeROM(string InFile, string FileName, BackgroundWorker worker)
...
+
+                byte[] ROM = ROMFuncs.BuildROM(FileName);
+                if (_outputVC)
+                {
+                    worker.ReportProgress(98, "Building VC...");
+                    ROMFuncs.BuildVC(ROM, VCDirectory, Path.ChangeExtension(FileName, "wad"));
+                }
             }
-            worker.ReportProgress(100, "Done!");
+                worker.ReportProgress(100, "Done!");
```
Comment (ZoeyZolotova) [inline Build.cs]: Why is this indented?

File: Shuffle.cs [hunk 4]
```diff
@@ -514,22 +516,98 @@ private void MakeSpoilerLog()
...
+        private void MakeHTMLSpoilerLog()
+        {
+            var settingsString = EncodeSettings();
+
+            var directory = Path.GetDirectoryName(Settings.OutputROMFilename);
+            var filename = $"{Path.GetFileNameWithoutExtension(Settings.OutputROMFilename)}_SpoilerLog.html";
+
+            StreamWriter LogFile = new StreamWriter(Path.Combine(directory, filename));
+
+            LogFile.WriteLine("<html>");
```
Comment (ZoeyZolotova) [inline Shuffle.cs]: I feel like the HTML should be in a Resource file. Probably don't need to go as far as using a compiler, but this should definitely be cleaned up.

File: fMain.cs [hunk 5]
```diff
@@ -276,6 +277,21 @@ private void cMixSongs_CheckedChanged(object sender, EventArgs e)
...
         }
 
+        private void cFreeHints_CheckedChanged(object sender, EventArgs e)
+        {
+            UpdateSingleSetting(() => Settings.FreeHints = cFreeHints.Checked);
+        }
+
+        private void cLogOnly_CheckedChanged(object sender, EventArgs e)
+        {
+            UpdateSingleSetting(() => Settings.FreeHints = cLogOnly.Checked);
```
Comment (ZoeyZolotova) [inline fMain.cs]: Shouldn't be FreeHints.

File: fMain.cs [hunk 6]
```diff
@@ -276,6 +277,21 @@ private void cMixSongs_CheckedChanged(object sender, EventArgs e)
...
+        }
+
+        private void cHMTLLog_CheckedChanged(object sender, EventArgs e)
+        {
+            UpdateSingleSetting(() => Settings.FreeHints = cHTMLLog.Checked);
```
Comment (ZoeyZolotova) [inline fMain.cs]: Shouldn't be FreeHints.

File: fMain.cs [hunk 7]
```diff
@@ -669,19 +666,31 @@ private void TryRandomize(BackgroundWorker worker, DoWorkEventArgs e)
             }
 
             // Additional validation of preconditions
-            if (!ValidateInputFile()) return;
-
-            if (!ValidateROM(Settings.InputROMFilename))
+            if (!cLogOnly.Checked)
```
Comment (ZoeyZolotova) [inline fMain.cs]: Looks really weird here. Should be
```
if (Settings.OutputROM)
{
    // validate input file
    // validate ROM (btw why are there two functions that seem like they do the same thing?)
    // MakeROM
}

if (Settings.OutputSpoilerLog)
{
    // WriteSpoilerLog
}

if (Settings.OutputHtmlSpoilerLog)
{
    // WriteHtmlSpoilerLog
}
```

File: Build.cs (hunk 3 above)
Comment (zmarotrix) [reply Build.cs]: Probably a mistake. I'll fix it.

File: fMain.cs (hunk 5 above)
Comment (zmarotrix) [reply fMain.cs]: lol whoops

File: Shuffle.cs (hunk 4 above)
Comment (zmarotrix) [reply Shuffle.cs]: I'm not sure exactly what you mean by this.

Comment (ZoeyZolotova) [reply Shuffle.cs]: i.e. have a resource that's like SpoilerLogHtmlTemplate
```<html><head><style>
...
<table ...>
<thead></thead>
<tbody>{0}</tbody>
<table>
...
</html>
```
and another that's like SpoilerLogTableRowTemplate
```
<tr><td>{0}</td><td class="spoiler">{1}</td></tr>
```
etc.

File: Build.cs [hunk 8]
```diff
@@ -462,13 +484,16 @@ private void MakeROM(string InFile, string FileName, BackgroundWorker worker)
             worker.ReportProgress(89, "Writing spoiler log...");
```
Comment (ZoeyZolotova) [inline Build.cs]: I think everything before this line should be wrapped in the `if (Settings.GenerateROM) {` if block. In fact maybe move WriteSpoilerLog to be the first thing in this function, then the rest is wrapped in GenerateROM

Comment (ZoeyZolotova) [reply Build.cs]: Actually move WriteSpoilerLog out of this function. It doesn't belong here. This is the MakeROM function, not the MakeEverything function.

File: fMain.cs [hunk 9]
```diff
@@ -107,6 +107,7 @@ private void bopen_Click(object sender, EventArgs e)
 
         private void bRandomise_Click(object sender, EventArgs e)
         {
+            if(Settings.GenerateROM)
```
Comment (ZoeyZolotova) [inline fMain.cs]:
```
if (Settings.GenerateROM && !ValidateInputFile())
{
    return;
}
```

File: fMain.cs [hunk 10]
```diff
@@ -276,6 +312,16 @@ private void cMixSongs_CheckedChanged(object sender, EventArgs e)
             UpdateSingleSetting(() => Settings.AddSongs = cMixSongs.Checked);
         }
 
+        private void cFreeHints_CheckedChanged(object sender, EventArgs e)
+        {
+            UpdateSingleSetting(() => Settings.FreeHints = cFreeHints.Checked);
+        }
+
+        private void cHMTLLog_CheckedChanged(object sender, EventArgs e)
```
Comment (ZoeyZolotova) [inline fMain.cs]: Why does this function exist twice? Line 239.

File: fMain.cs [hunk 11]
```diff
@@ -669,19 +700,32 @@ private void TryRandomize(BackgroundWorker worker, DoWorkEventArgs e)
...
 
-            MessageBox.Show("Successfully built output ROM!",
-                "Success", MessageBoxButtons.OK, MessageBoxIcon.None);
+            if (Settings.GenerateROM)
+            {
+                MakeROM(Settings.InputROMFilename, Settings.OutputROMFilename, worker);
+                MessageBox.Show("Successfully built output ROM!",
+                    "Success", MessageBoxButtons.OK, MessageBoxIcon.None);
+            }
+            else
```
Comment (ZoeyZolotova) [inline fMain.cs]: As i mentioned in another comment, this shouldn't be `else` it should be `if (Settings.GenerateSpoilerLog)` and remove WriteSpoilerLog from the MakeROM function.

File: Build.cs (hunk 8 above)
Comment (zmarotrix) [reply Build.cs]: I was wondering about that too. I'll ask Kjelli about it since he wrote this bit.

File: Build.cs [hunk 12]
```diff
@@ -206,6 +206,16 @@ private void WriteEnemies()
                 SeedRNG();
                 ROMFuncs.ShuffleEnemies(RNG);
             }
+        }
+
+        /// <summary>
+        /// Update the gossip stone actor to not check mask of truth
+        /// </summary>
+        private void WriteFreeHints()
```
Comment (ZoeyZolotova) [inline Build.cs]: Overall I do like this code better than apply a hack file that does who-knows-what. hack file contents would be `00 E0 AB 88 01 00 FF`

File: Build.cs [hunk 13]
```diff
@@ -412,63 +444,64 @@ private bool ValidateROM(string FileName)
 
         private void MakeROM(string InFile, string FileName, BackgroundWorker worker)
         {
+            if(Settings.GenerateROM)
```
Comment (ZoeyZolotova) [inline Build.cs]: This line is unnecessary. It should be (and already is) checked where this function is called.

File: Shuffle.cs [hunk 14]
```diff
@@ -475,61 +475,52 @@ private void SetTatlColour()
             };
         }
 
-        private void MakeSpoilerLog()
+        private void CreateTextSpoilerLog(Spoiler spoiler, string path)
```
Comment (ZoeyZolotova) [inline Shuffle.cs]: I would like the text spoiler log to also be done using a template. Something to do later i suppose.

**Pull request merged**